File: src/llm/googleTrends.js
// src/llm/googleTrends.js
// Minimal, dependency-free Google Trends client (no API key, no external lib).
//
// Endpoint: https://trends.google.com/trending/rss?geo=XX  — Google's public
// "Daily Search Trends" RSS feed. Gives a snapshot of what people are actually
// searching right now per region (US/NG/GB...). We use it as an INSPIRATION /
// "human pulse" layer: the V10 script picks its psychological angle from it and
// the Shorts seed fresh vocabulary from it — never as a hard directive.
//
// Rules:
//   - BEST EFFORT ONLY: any failure returns [] so no pipeline ever breaks on
//     Google being unreachable/rate-limited.
//   - CACHED (~3h) so the daily build does not hammer the feed.
//   - No scraping of autocomplete/cookies — just the public RSS.
const fs = require("fs");
const path = require("path");

const BASE = process.env.FORCHI_BASE || path.resolve(__dirname, "..", "..");
const CACHE_FILE = path.join(BASE, "temp_media", "google_trends_cache.json");
const CACHE_TTL_MS = 3 * 60 * 60 * 1000; // 3 hours
const RSS_URL = "https://trends.google.com/trending/rss";
const UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120 Safari/537.36";

function readCache() {
  try {
    const c = JSON.parse(fs.readFileSync(CACHE_FILE, "utf8"));
    if (c && c.at && Date.now() - c.at < CACHE_TTL_MS && Array.isArray(c.terms)) return c.terms;
  } catch { /* first run / corrupt */ }
  return null;
}
function writeCache(terms) {
  try { fs.writeFileSync(CACHE_FILE, JSON.stringify({ at: Date.now(), terms }, null, 1)); } catch {}
}

function decodeXml(s) {
  return String(s || "")
    .replace(/<!\[CDATA\[/g, "").replace(/\]\]>/g, "")
    .replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&apos;/g, "'")
    .replace(/&#(\d+);/g, (_, n) => { try { return String.fromCodePoint(+n); } catch { return ""; } })
    .trim();
}

// Split the RSS into <item>...</item> blocks and pull each item's <title>.
function parseRss(xml) {
  const items = [];
  const re = /<item>([\s\S]*?)<\/item>/g;
  let m;
  while ((m = re.exec(xml)) !== null) {
    const body = m[1];
    const t = decodeXml((body.match(/<title>([\s\S]*?)<\/title>/) || [])[1] || "");
    const clean = t.replace(/\s+/g, " ").trim();
    if (!clean || clean.toLowerCase() === "daily search trends") continue;
    // Keep terms a spark-writer can actually use: drop heavy non-Latin runs
    // (CJK/Greek/Cyrillic) while keeping accented Latin + digits.
    const letters = clean.replace(/[^A-Za-zÀ-ÿ]/g, "").length || 1;
    const nonAscii = clean.replace(/[\x00-\x7F]/g, "").length;
    if (nonAscii / letters > 0.25) continue;
    items.push(clean);
  }
  return items;
}

async function fetchGeo(geo) {
  const url = `${RSS_URL}?geo=${encodeURIComponent(geo)}`;
  const ctrl = new AbortController();
  const to = setTimeout(() => ctrl.abort(), 20000);
  try {
    const r = await fetch(url, { headers: { "User-Agent": UA, "Accept": "application/rss+xml, text/xml, */*" }, signal: ctrl.signal });
    if (!r.ok) { console.warn(`[googleTrends] geo=${geo} HTTP ${r.status}`); return []; }
    const xml = await r.text();
    const terms = parseRss(xml);
    return terms.slice(0, 25);
  } catch (e) {
    console.warn(`[googleTrends] geo=${geo} failed:`, e.message);
    return [];
  } finally {
    clearTimeout(to);
  }
}

// Combined today's searches across several geos, deduped, order preserved.
async function getTrending({ geos = ["US", "NG", "GB"], maxPerGeo = 15, maxTotal = 35, force = false } = {}) {
  if (!force) {
    const cached = readCache();
    if (cached && cached.length) return cached.slice(0, maxTotal);
  }
  const settled = await Promise.allSettled(geos.map((g) => fetchGeo(g)));
  const seen = new Set();
  const out = [];
  for (const res of settled) {
    if (res.status !== "fulfilled") continue;
    for (const term of res.value.slice(0, maxPerGeo)) {
      const key = term.toLowerCase();
      if (seen.has(key)) continue;
      seen.add(key);
      out.push(term);
      if (out.length >= maxTotal) break;
    }
    if (out.length >= maxTotal) break;
  }
  if (out.length) writeCache(out);
  return out;
}

// Alias used by script generators: plain list of strings (cached).
async function trendingTerms(opts) {
  return getTrending(opts);
}

module.exports = { getTrending, trendingTerms, fetchGeo, parseRss };

if (require.main === module) {
  (async () => {
    const terms = await getTrending({ force: true });
    console.log("trending terms:", JSON.stringify(terms, null, 1));
  })();
}

File: tools/_google_trends.py
"""tools/_google_trends.py — dependency-free Google Trends "Daily Search Trends"
sparks for the Python side of the pipeline (Shorts). Mirrors src/llm/googleTrends.js.

Endpoint: https://trends.google.com/trending/rss?geo=XX (public RSS, no key).
BEST EFFORT ONLY: any failure returns [] so the Shorts pipeline never breaks on
Google being unreachable. Cached ~3h in temp_media/google_trends_sparks.json.

Usage:
    python tools/_google_trends.py            # print fresh sparks
    from _google_trends import fetch_trend_sparks
"""
import json
import os
import re
import sys
import time
import urllib.parse
import urllib.request
import xml.etree.ElementTree as ET

try:
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
except Exception:
    pass

try:
    from _paths import BASE
except Exception:
    BASE = os.path.abspath(os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

RSS_URL = "https://trends.google.com/trending/rss"
CACHE_PATH = os.path.join(BASE, "temp_media", "google_trends_sparks.json")
CACHE_TTL = 3 * 60 * 60  # 3h
GEO_DEFAULT = ("US", "NG", "GB")
UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120 Safari/537.36"


def _read_cache():
    try:
        with open(CACHE_PATH, encoding="utf-8") as f:
            c = json.load(f)
        if c and isinstance(c.get("terms"), list) and time.time() - c.get("at", 0) < CACHE_TTL:
            return c["terms"]
    except Exception:
        pass
    return None


def _write_cache(terms):
    try:
        with open(CACHE_PATH, "w", encoding="utf-8") as f:
            json.dump({"at": time.time(), "terms": terms}, f, ensure_ascii=False)
    except Exception:
        pass


def _clean(term):
    term = re.sub(r"<!\[CDATA\[|\]\]>", "", term or "")
    term = (term.replace("&amp;", "&").replace("&lt;", "<").replace("&gt;", ">")
                .replace("&quot;", '"').replace("&#39;", "'").replace("&apos;", "'"))
    term = re.sub(r"\s+", " ", term).strip()
    if not term or term.lower() == "daily search trends":
        return None
    letters = len(re.sub(r"[^A-Za-z\u00C0-\u00FF]", "", term)) or 1
    non_ascii = len(re.sub(r"[\x00-\x7f]", "", term))
    if non_ascii / letters > 0.25:
        return None
    return term


def _fetch_geo(geo):
    url = "%s?geo=%s" % (RSS_URL, urllib.parse.quote(geo))
    req = urllib.request.Request(url, headers={"User-Agent": UA, "Accept": "application/rss+xml, text/xml, */*"})
    try:
        with urllib.request.urlopen(req, timeout=20) as r:
            data = r.read().decode("utf-8", "replace")
        root = ET.fromstring(data)
        terms = []
        for item in root.iter("item"):
            title = item.findtext("title")
            c = _clean(title)
            if c:
                terms.append(c)
        return terms[:25]
    except Exception as e:
        print("[googleTrends] geo=%s failed: %s" % (geo, str(e)[:80]), file=sys.stderr)
        return []


def fetch_trend_sparks(geos=GEO_DEFAULT, max_per_geo=15, max_total=35, force=False):
    """Return today's trending search terms across the given geos (best-effort)."""
    if not force:
        cached = _read_cache()
        if cached:
            return cached[:max_total]
    seen, out = set(), []
    for geo in geos:
        for term in _fetch_geo(geo):
            key = term.lower()
            if key in seen:
                continue
            seen.add(key)
            out.append(term)
            if len(out) >= max_per_geo:
                break
        if len(out) >= max_total:
            break
    if out:
        _write_cache(out)
    return out


if __name__ == "__main__":
    sparks = fetch_trend_sparks(force=True)
    print("sparks (%d):" % len(sparks))
    for s in sparks:
        print("-", s)
